src/include/random.h

```diff
@@ -0,0 +1,391 @@
+// -*- mode:C++; tab-width:8; c-basic-offset:2; indent-tabs-mode:t -*-
+// vim: ts=8 sw=2 smarttab
+/*
+ * Ceph - scalable distributed file system
+ *
+ * Copyright (C) 2017 SUSE LINUX GmbH
+ *
+ * This is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License version 2.1, as published by the Free Software
+ * Foundation.  See file COPYING.
+ *
+*/
+
+#ifndef CEPH_RANDOM_H
+#define CEPH_RANDOM_H 1
+
+#include <mutex>
+#include <random>
+
+// Basic random number facility, adapted from N3551:
+namespace ceph {
+namespace util {
+
+inline namespace version_2017_0 {
+
+namespace detail {
+
+template <typename T>
+T& urng()
+{
+ static T rng_engine;
+ return rng_engine;
+}
+
+template <typename MutexT, typename EngineT>
+void randomize_rng(const int seed, MutexT& m, EngineT& e)
+{
+ std::lock_guard<MutexT> lg(m);
+ e.seed(seed);
+}
+
+template <typename MutexT, typename EngineT>
+void randomize_rng(MutexT& m, EngineT& e)
+{
+ static std::random_device rd;
+
+ std::lock_guard<MutexT> lg(m);
+ e.seed(rd());
+}
+
+template <typename MutexT, 
+          typename EngineT = std::default_random_engine>
+void randomize_rng(MutexT& m)
+{
+ randomize_rng(m, detail::urng<EngineT>());
+}
+
+template <typename MutexT, 
+          typename EngineT = std::default_random_engine>
+void randomize_rng(const int n, MutexT& m)
+{
+ std::lock_guard<MutexT> lg(m);
+ detail::urng<EngineT>().seed(n);
+}
+
```

Comment:
Hi! Thanks for your patience while I got the gist of this-- here are some non-random thoughts. 
It looks like the purpose of this is to improve the quality of the random number seeding. I think that might be a good idea, but also that it should come as a follow up PR. 

The change at hand is meant to provide an alternative to rand(), srand(), and
mod, while hiding some of the details of <random> from the user (aka "easy
to use"). The default random engine is "supposedly" good enough for this  
purpose, given that here there's no attempt to generate cryptograpically 
strong random numbers.

I'm definitely aware that there are some criticisms of several aspects of
<random>, and another hopeful side-effect is that when something like     
N4316
("http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2014/n4316.html") is
adopted, we will easily be able to change to using it with everything
being centralized. (As for deeper criticism of <random>, I'm going to stay
out of that one-- it might be worthwhile to send proposals to the
committee, or talk with someone involved.)

It might even be possible to hook seeding improvements like your suggestion shows into this randomization
mini-library via the engine hooks in this PR, which lets the
user choose and would avoid affect any code. (Assuming it's not best to nip in the bud now.)

Ultimately, this is a nerfed version of <random>: if we
get /too/ general, or the user has a very specific or expert-domain
problem, they probably should use <random> directly, or the tools of their
choice. I'm also ok with replacing the RNG seeding if folks feel strongly
about it and think it's going to cover the cases needed. My goal here is  
to do better than rand() and deliver thread safety and /some/ flexibility.

I wouldn't want to see the baby thrown out with the bathwater-- certainly,
since it sounds like there's utility in providing stronger seeding it
should be considered-- I think a PR based on your extension would be valuable.